src/main.rs

```diff
@@ -37,14 +38,45 @@ struct MarketData {
 }
 
 impl MarketData {
-    fn print_holders_map(&self) {
-        for (hash, holders) in &self.files {
+    fn validate_holders(&self, hash: &String) -> Vec<FileRequest> {
+        // make a map of holders already printed
+        let mut previous_holders: HashMap<&String, &FileRequest> = HashMap::new();
+
+        for holder in &self.files[hash] {
+            let user = &holder.user;
+            if previous_holders.contains_key(&user.id) {
+                // check which holder has the most recent ttl
+                // if the current holder has a more recent ttl, print it and remove the previous holder
+                // if the previous holder has a more recent ttl, skip the current holder
+                let prev_holder = previous_holders.get(&user.id).unwrap();
+                let current_holder_ttl = holder.expiration - get_current_time(); // current holder: smaller means older
+                let previous_holder_ttl: u64 = &prev_holder.expiration - get_current_time(); // previous holder: smaller means older
+
+                if current_holder_ttl > previous_holder_ttl { // if the current holder is younger, remove the previous holder
+                    previous_holders.insert(&user.id, &holder);
+                }
+                continue;
+            }
+            previous_holders.insert(&user.id, &holder);
+        }
+        return previous_holders.into_iter().map(|(_, holder)| holder.clone()).collect();
+    }
+
+    fn print_holders_map(&mut self) {
+        let mut validated_files: HashMap<String, Vec<FileRequest>> = HashMap::new();
+
+        for (hash, _) in &self.files {
             println!("File Hash: {hash}");
-            for holder in holders {
+            let validated_holders = self.validate_holders(&hash);
```

Comment:
I was thinking it would be more efficient to handle this in Register File. When insterting a new user, check if they already exist and also do the TTL garbage collection, and insert the new user at the end of the list. That way, there won't be any duplicate users in this query. Here is an issue I recently made that decsribes this (https://github.com/sbu-416-24sp/orcanet-market-rust/issues/4)

Reply:
Good idea, I've updated the function to validate the holders every time a new producer is registered for that specific file hash, and I've also tested this with the peer node code to confirm that it works.

Not sure how to link an issue to this pr, but it should resolve the issue you mentioned